plugins/filter_cgolib/cgolib.c

```diff
@@ -0,0 +1,276 @@
+        msgpack_pack_array(&tmp_pck, 2);
+        flb_time_append_to_msgpack(&tm, &tmp_pck, 0);
+
+        map_num = obj->via.map.size;
+
+        // add tag to slice
+        name[0].p = "tag";
+        name[0].n = 3;
+        value[0].p = tag;
+        value[0].n = tag_len;
+        index = 1;
+        for (i=0;i<map_num; i++) {
+            kv = &obj->via.map.ptr[i];
+            if ( msgpackobj2char(&kv->key, &key_str, &key_len) < 0 ) {
+                /* key is not string */
+                continue;
+            }
+            if ( msgpackobj2char(&kv->val, &val_str, &val_len) < 0 ) {
+                /* val is not string */
+                continue;
+            }
+            name[index].p = key_str;
+            name[index].n = key_len;
+            value[index].p= val_str;
+            value[index].n = val_len;
+            index++;
+        }
+
+        name_slice.len = index;
+        value_slice.len = index;
+        // call cgo filter
+        ret = ctx->filter_lib_func(name_slice, value_slice);
+        if (ret == -1) {
+            flb_error("cgolib filter_lib_func fail. ");
+            msgpack_unpacked_destroy(&unpackMsg);
+            return -1;
+        }
+
+        // load result
+        msgpack_pack_map(&tmp_pck, ret);
+        for (i=0;i<ret;i++) {
+            msgpack_pack_str(&tmp_pck, name[i].n);
+            msgpack_pack_str_body(&tmp_pck, name[i].p, name[i].n);
+
+            msgpack_pack_str(&tmp_pck, value[i].n);
+            msgpack_pack_str_body(&tmp_pck, value[i].p, value[i].n);
+        }
+    }
+    msgpack_unpacked_destroy(&unpackMsg);
+
+    return 0;
+}
+
+
+static int cb_golib_init(struct flb_filter_instance *f_ins,
+                        struct flb_config *config,
+                        void *data)
+{
+    struct cgolib_conf *ctx;
+
+    char *error;
+
+    struct mk_list *head;
+    struct flb_config_prop *p;
+
+    char lib_so_name[200];
+
+    GoInt ret;
+    GoSlice name_slice, value_slice; // for []string  slice
+    _GoString_ name[MAX_PARAMETERS], value[MAX_PARAMETERS];
+    int para_num; // actual parametres num
+
+    ctx = malloc(sizeof(struct cgolib_conf));
+    f_ins->context = ctx;
+
+    name_slice.data = name;
+    value_slice.data = value;
+
+    para_num = 0;
+    mk_list_foreach(head, &f_ins->properties) {
+        p = mk_list_entry(head, struct flb_config_prop, _head);
+        if (strcasecmp(GOLIB_SO_KEY_NAME, p->key) == 0) {
+            strcpy(lib_so_name, p->val);
+            continue;
+        }
+        name[para_num].p = p->key;
+        name[para_num].n = strlen(p->key);
+
+        value[para_num].p = p->val;
+        value[para_num].n = strlen(p->val);
+
+        para_num++;
+    }
+    name_slice.len = para_num;
+    name_slice.cap = para_num;
+    value_slice.len = para_num;
+    value_slice.cap = para_num;
+
+    ctx->handler = dlopen (lib_so_name, RTLD_LAZY);
+    if (!ctx->handler) {
+        flb_error("open lib so %s [%s] fail. ", lib_so_name, dlerror());
```

Comment:
Format the errors so that it indicates the source of the error. For example,
`[filter_cgolib] open lib so fail`

Reply:
ok, I will do that